packages/fcl/src/wallet-provider-spec/draft-v3.md

```diff
@@ -231,7 +198,299 @@ const reason = "User declined to authenticate."
 WalletUtils.decline(reason)
 ```
 
-### `data` and `params`
+### `Service`
+
+```typescript
+type ServiceType =
+  | 'authn'
+  | 'authz'
+  | 'user-signature'
+  | 'pre-authz'
+  | 'open-id'
+  | 'back-channel-rpc'
+  | 'authn-refresh'
+
+type ServiceMethod =
+  | 'HTTP/POST'
+  | 'IFRAME/RPC'
+  | 'POP/RPC'
+  | 'TAB/RPC'
+  | 'EXT/RPC'
+  | 'DATA'
+
+interface Service extends ObjectBase {
+  f_type: 'Service'
+  type: ServiceType
+  method: ServiceMethod
+  uid: string
+  endpoint: string
+  id: string
+  identity: Identity
+  provider?: ServiceProvider
+  data?: FclObject
+}
+```
+
+The meaning of the fields is as follows.
+- `type`: The type of this service.
+- `method`: The service method this service uses. `DATA` means that the purpose of this service is just to provide the information in this `Service` object, and no active communication services are provided.
+- `uid`: A unique identifier for the service. A common scheme for deriving this is to use `'wallet-name#${type}'`, where `${type}` refers to the type of this service.
+- `endpoint`: Defines where to communicate with the service.
+  - When `method` is `EXT/RPC`, this can be an arbitrary unique string, and the extension will need to use it to identify its own services. A common scheme for deriving the `endpoint` is to use `'ext:${address}'`, where `${address}` refers to the wallet's address. (See `ServiceProvider` for more information.)
+- `id`: The wallet's internal identifier for the user. If no other identifier is used, simply the user's flow account address can be used here.
+- `identity`: Information about the identity of the user.
+- `provider`: Information about the wallet.
+- `data`: Additional information used with a service of type `open-id`.
+
+See also:
+- [authn](https://github.com/onflow/flow-js-sdk/blob/master/packages/fcl/src/current-user/normalize/authn.js)
+- [authz](https://github.com/onflow/flow-js-sdk/blob/master/packages/fcl/src/current-user/normalize/authz.js)
+- [user-signature](https://github.com/onflow/flow-js-sdk/blob/master/packages/fcl/src/current-user/normalize/user-signature.js)
+- [pre-authz](https://github.com/onflow/flow-js-sdk/blob/master/packages/fcl/src/current-user/normalize/pre-authz.js)
+- [open-id](https://github.com/onflow/flow-js-sdk/blob/master/packages/fcl/src/current-user/normalize/open-id.js)
+- [back-channel-rpc](https://github.com/onflow/flow-js-sdk/blob/master/packages/fcl/src/current-user/normalize/back-channel-rpc.js)
+
+### `Identity`
+This object is used to define the identity of the user.
+
+```typescript
+interface Identity extends ObjectBase {
+  f_type: 'Identity'
+  address: string
+  keyId?: number
+}
+```
+
+The meaning of the fields is as follows.
+- `address`: The flow account address of the user.
+- `keyId`: The user's key id.
```

Comment:
Please suggest a better description if you can, I don't quite get what `keyId` is for...

Reply:
Thanks, I made the description more precise with this understanding.